File: 117-populating-next-right-pointers-in-each-node-ii/117-populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if(!root)
            return NULL;
        queue<Node*> q;
        q.push(root);
        while(!q.empty())
        {
            int sz = q.size();
            while(sz--)
            {
                Node* f = q.front();
                q.pop();
                
               
                if(f->left)
                    q.push(f->left);
                if(f->right)
                    q.push(f->right);
                
                 if(sz==0)
                    f->next= NULL;
                else
                    f->next = q.front();
            }
            //level is exhausted
            
        }
        
        return root;
    }
};

File: 117-populating-next-right-pointers-in-each-node-ii/117-populating-next-right-pointers-in-each-node-ii.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var left: Node? = null
 *     var right: Node? = null
 *     var next: Node? = null
 * }
 */

class Solution {
    fun connect(root: Node?, level: Int = 0, levels: ArrayList<Node?> = ArrayList()): Node? {
    if(root != null) {
        if(levels.size <= level) levels.add(root) else {
            levels[level]?.next = root
            levels[level] = root
        }
        connect(root.left, level + 1, levels)
        connect(root.right, level + 1, levels)
    }
    return root
}
}